test/test_buf_error.c

```diff
@@ -0,0 +1,303 @@
+#include "test.h"
+#include "test_utils.h"
+
+#define MAX_LOOPS 20
+
+/*
+ * Here there is a small collection of test cases for when inflate must return Z_BUF_ERROR.
+ * 1. test_0_avail_in - no progress can be made because avail_in is 0.
+ * 2. test_small_avail_out tests - no progress can be made because avail_out is not enough
+ * to finish at once.
+ * 3. test_scp - replicates an issue found using libnxz with scp.
+ */
+
+int test_0_avail_in() {
+
+        z_stream strm = {0};
+        Byte *compr = NULL, *uncompr = NULL;
+        unsigned int outlen = 4096;
+        int rc = 0, ret = 0;
+
+        if ((rc = inflateInit(&strm)) != 0){
+		printf("*** failed: inflateInt returned %d\n", rc);
+		return TEST_ERROR;
+	}
+
+        compr = (Byte*)calloc(1,1);
+	if ( compr == NULL ) {
+		printf("*** failed: failed to allocate memory");
+		return TEST_ERROR;
+	}
+
+        uncompr = (Byte*)calloc(outlen, sizeof(Byte));
+        if ( uncompr == NULL ) {
+		free(compr);
+		printf("*** failed: failed to allocate memory");
+		return TEST_ERROR;
+	}
+
+        strm.avail_out = outlen;
+        strm.next_out = uncompr;
+        strm.next_in = compr;
+
+	/* At this point avail_in is 0 */
```

Comment:
Between having this comment and a line with `strm.avail_in = 0` , I believe the second option is better. It's just as clear and also protects against a possible breakage if someone modifies the struct initialization inadvertently.